File: backend/internal/model/material.go
package model

import "time"

type LearningMaterial struct {
	ID              string    `db:"id" json:"id"`
	UserID          string    `db:"user_id" json:"-"`
	LearningSpaceID string    `db:"learning_space_id" json:"learning_space_id"`
	MaterialKind    string    `db:"material_kind" json:"material_kind"`
	OriginalName    string    `db:"original_name" json:"original_name"`
	ObjectKey       string    `db:"object_key" json:"-"`
	MIMEType        string    `db:"mime_type" json:"mime_type"`
	SizeBytes       int64     `db:"size_bytes" json:"size_bytes"`
	ContentSHA256   []byte    `db:"content_sha256" json:"-"`
	Status          string    `db:"status" json:"status"`
	FailureReason   *string   `db:"failure_reason" json:"failure_reason"`
	CreatedAt       time.Time `db:"created_at" json:"created_at"`
	UpdatedAt       time.Time `db:"updated_at" json:"updated_at"`
}

type GenerationJob struct {
	ID              string     `db:"id" json:"id"`
	UserID          string     `db:"user_id" json:"-"`
	LearningSpaceID string     `db:"learning_space_id" json:"learning_space_id"`
	MaterialID      string     `db:"material_id" json:"material_id"`
	JobType         string     `db:"job_type" json:"job_type"`
	Status          string     `db:"status" json:"status"`
	Progress        int        `db:"progress" json:"progress"`
	Attempts        int        `db:"attempts" json:"attempts"`
	MaxAttempts     int        `db:"max_attempts" json:"max_attempts"`
	AvailableAt     time.Time  `db:"available_at" json:"-"`
	StartedAt       *time.Time `db:"started_at" json:"started_at"`
	CompletedAt     *time.Time `db:"completed_at" json:"completed_at"`
	ErrorCode       *string    `db:"error_code" json:"error_code"`
	ErrorMessage    *string    `db:"error_message" json:"error_message"`
	CreatedAt       time.Time  `db:"created_at" json:"created_at"`
	UpdatedAt       time.Time  `db:"updated_at" json:"updated_at"`
}

File: frontend/src/types/material.ts
export type MaterialStatus = 'queued' | 'processing' | 'ready' | 'failed'
export type JobStatus = 'queued' | 'processing' | 'succeeded' | 'failed'

export interface LearningMaterial {
  id: string
  learning_space_id: string
  material_kind: 'study' | 'past_exam'
  original_name: string
  mime_type: string
  size_bytes: number
  status: MaterialStatus
  failure_reason: string | null
  created_at: string
  updated_at: string
}

export interface GenerationJob {
  id: string
  learning_space_id: string
  material_id: string
  job_type: 'process_material' | 'embed_material'
  status: JobStatus
  progress: number
  attempts: number
  max_attempts: number
  started_at: string | null
  completed_at: string | null
  error_code: string | null
  error_message: string | null
  created_at: string
  updated_at: string
}

export interface MaterialListItem {
  material: LearningMaterial
  job: GenerationJob
  chunk_count: number
}

export interface MaterialList {
  items: MaterialListItem[]
}

export type MaterialUploadResult = MaterialListItem

export interface MaterialChunk {
  id: string
  material_id: string
  chunk_index: number
  content: string
  source_type: 'page' | 'slide' | 'document' | 'text'
  source_start: number | null
  source_end: number | null
  char_count: number
  token_estimate: number
  created_at: string
}

export type MaterialIndexState =
  | 'not_started'
  | 'pending'
  | 'indexing'
  | 'indexed'
  | 'partial'
  | 'failed'

export interface MaterialIndexStatus {
  material_id: string
  status: MaterialIndexState
  embedding_model: string
  total_chunks: number
  pending_chunks: number
  indexed_chunks: number
  failed_chunks: number
  progress: number
  job: GenerationJob | null
}

export interface MaterialDetail {
  material: LearningMaterial
  processing_job: GenerationJob
  embedding_job: GenerationJob | null
  chunk_count: number
}

export interface MaterialDownload {
  url: string
  expires_at: string
}
